prow/config/secrets_agent.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Implements an agent to read and reload the secrets.
+package config
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// SecretAgent watches a path and automatically loads the secrets stored.
+type SecretAgent struct {
+	sync.Mutex
+	secretsMap map[string][]byte
+}
+
+// NewSecretAgent returns a SecretAgent with a map of strings with []byte value.
+// The key will be used to index the file path, and the value to store the secret value.
+func NewSecretAgent() *SecretAgent {
+	return &SecretAgent{
+		secretsMap: make(map[string][]byte),
+	}
+}
+
+// Start will begin polling the secret file at the path. If the first load
+// fails, Start with return the error and abort. Future load failures will log
+// the failure message but continue attempting to load.
+func (sa *SecretAgent) Start(paths []string) error {
+	secretsMap, err := LoadSecrets(paths)
+	if err != nil {
+		return err
+	}
+
+	sa.secretsMap = secretsMap
+
+	go func() {
+		var lastModTime time.Time
+
+		skips := 0
+		for range time.Tick(1 * time.Second) {
+			for secretPath := range secretsMap {
+				if skips < 600 {
+					// Check if the file changed to see if it needs to be re-read.
+					secretStat, err := os.Stat(secretPath)
+					if err != nil {
+						logrus.WithField("secretPath", secretPath).WithError(err).Error("Error loading secret file.")
+						continue
+					}
+
+					recentModTime := secretStat.ModTime()
+
+					if !recentModTime.After(lastModTime) {
+						skips++
+						continue // file hasn't been modified
+					}
+					lastModTime = recentModTime
+				}
+
+				if secretsMap, err := LoadSecrets(paths); err != nil {
+					logrus.WithField("Secret paths", paths).
+						WithError(err).Error("Error loading config.")
```

Comment:
Trying to load all of the secrets when we find that one secret has updated doesn't seem like the right approach, especially if we continue on to check the other secrets immediately after reloading them all.
Additionally, with this pattern, if one secret fails to update we will fail to update all other secrets as well.